cloudmarker/test/test_esstore.py

```diff
@@ -0,0 +1,73 @@
+"""Tests for EsStore plugin."""
+
+import unittest
+from unittest import mock
+
+from elasticsearch import ElasticsearchException
+
+from cloudmarker.stores import esstore
+
+
+class EsStoreTest(unittest.TestCase):
+    """Tests for EsStore plugin."""
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_happy_flow(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk_response = {
+            'errors': '',
+            'items': [
+                {'index': {'status': 200}}
+            ]
+        }
+
+        mock_bulk.return_value = mock_bulk_response
+
+        es_store.write(mock_record)
+        es_store.done()
```

Comment:
After the `write`, I'd check if the buffer actually got filled though that's up to your discretion if you want it here or not. This test works well and asserting the buffer after the write is just for more clarity.

Reply:
I doubt there is a way to check if the buffer got filled after `write()` without relying on testing the internal state. Asserting on an internal attribute breaks encapsulation and tightly couples the test methods with the internal implementation of a class, so it should be avoided. I discussed this in some detail [here](https://github.com/cloudmarker/cloudmarker/pull/96#discussion_r277122687).

What we should test instead is whether the public methods of Elasticsearch were invoked like we expect with the help of mocks. This has been done below in line 32, so I think this test method is good.

One might argue how Elasticsearch is used in a class is also an internal implementation detail. The answer lies in the publicly visible behaviour we want to test. We do not want to test how `EsStore` internally manages its buffer. We only to test that if `EsStore`'s `write()` and `done()` methods are called, then we hand over the records to Elasticsearch. To test this it becomes necessary to at least ensure that `Elasticsearch`'s `bulk()` method was called. That's the maximum extent of how far we should go regarding testing internal implementation of a class.